src/process/index/Indexer.h

```diff
@@ -0,0 +1,202 @@
+     * Indexer version available.
+     * If the Indexer ever changes, keep in mind to increment this version! This will be used to select the appropriate
+     * Extractor class when an index is read!
+     */
+    static const unsigned int INDEXER_VERSION;
+
+private:
+
+    long numberOfFoundEntries = 0;
+
+    shared_ptr<IndexWriter> indexWriter;
+
+    bool forbidWriteFQI;
+
+    bool forceOverwrite{false};
+
+    shared_ptr<IndexEntryStorageStrategy> storageStrategy;
+
+    bool compressDictionaries{true};
+
+    /**
+     * For debug and test purposes, used when debuggingEnabled is true
+     * keeps the index header
+     */
+    shared_ptr<IndexHeader> storedHeader = shared_ptr<IndexHeader>(nullptr);
+
+    /**
+     * For debug and test purposes, used when debuggingEnabled is true
+     * Keeps all generated index entries
+     */
+    vector<shared_ptr<IndexEntryV1>> storedEntries;
+
+    /**
+     * For debug and test purposes, used when decompressed blocks are processed. Will store them
+     * to a file next to the
+     */
+    bool writeOutOfDecompressedBlocksAndStatistics{false};
+
+    path storageForDecompressedBlocks;
+
+    bool writeOutOfPartialDecompressedBlocks{false};
+
+    path storageForPartialDecompressedBlocks;
+
+    ofstream partialBlockinfoStream;
+
+    /**
+     * Current bits for the next index entry.
+     */
+    int curBits{0};
+
+    /**
+     * Current offset in file.
+     */
+    long offset{0};
+
+    /**
+     * Marks, if the last inflated block ended with the newline character.
+     * If true, the blockOffsetInRawFile for the first line in the new block will be 0.
+     * If false, we have to look for the byte blockOffsetInRawFile.
+     */
+    bool lastBlockEndedWithNewline = true;
+
+    int64_t lineCountForNextIndexEntry{0};
+
+    int64_t numberOfConcatenatedFiles{1};
+
+    long blockID{-1};                   // Number of the currently processed block.
+
+    shared_ptr<IndexEntryV1> lastStoredEntry;
+
+    /**
+     * When we write out a new index entry, we need a dictionary. This is taken from the block of data before the block
+     * to which the new entry references.
+     */
+    Bytef dictionaryForNextBlock[WINDOW_SIZE]{0};
+
+public:
+
+
+    /**
+     * Be careful, when you enableDebugging. This will tell the Indexer to store information about the process, which can
+     * e.g. be used for unit test. E.g. this will store ALL lines found in the FASTQ file! So it is absolutely not
+     * advisable to use it with large FASTQ files. For test data it is safe to use.
+     * @param fastq The FASTQ we are working on.
+     * @param index The index for the FASTQ.
+     * @param enableDebugging Store debug information or not.
+     */
+    Indexer(const shared_ptr<Source> &fastqfile,
+            const shared_ptr<Sink> &index,
+            const shared_ptr<IndexEntryStorageStrategy> &storageStrategy,
+            bool enableDebugging = false,
```

Comment:
If this parameter is of lower priority than the three below, consider reordering the parameters by order of expected usage frequency (most frequent first).